libs/dyn/dynvar/resolve.go

```diff
@@ -0,0 +1,203 @@
+package dynvar
+
+import (
+	"errors"
+	"fmt"
+	"slices"
+	"sort"
+	"strings"
+
+	"github.com/databricks/cli/libs/dyn"
+	"golang.org/x/exp/maps"
+)
+
+// Resolve resolves variable references in the given input value using the provided lookup function.
+// It returns the resolved output value and any error encountered during the resolution process.
+//
+// For example, given the input value:
+//
+//	{
+//	    "a": "a",
+//	    "b": "${a}",
+//	    "c": "${b}${b}",
+//	}
+//
+// The output value will be:
+//
+//	{
+//	    "a": "a",
+//	    "b": "a",
+//	    "c": "aa",
+//	}
+//
+// If the input value contains a variable reference that cannot be resolved, an error is returned.
+// If a cycle is detected in the variable references, an error is returned.
+// If for some path the resolution function returns [ErrSkipResolution], the variable reference is left in place.
+// This is useful when some variable references are not yet ready to be interpolated.
+func Resolve(in dyn.Value, fn Lookup) (out dyn.Value, err error) {
+	return resolver{in: in, fn: fn}.run()
+}
+
+type resolver struct {
+	in dyn.Value
+	fn Lookup
+
+	refs     map[string]ref
+	resolved map[string]dyn.Value
+}
+
+func (r resolver) run() (out dyn.Value, err error) {
+	err = r.collectVariableReferences()
+	if err != nil {
+		return dyn.InvalidValue, err
+	}
+
+	err = r.resolveVariableReferences()
+	if err != nil {
+		return dyn.InvalidValue, err
+	}
+
+	out, err = r.replaceVariableReferences()
+	if err != nil {
+		return dyn.InvalidValue, err
+	}
+
+	return out, nil
+}
+
+func (r *resolver) collectVariableReferences() (err error) {
+	r.refs = make(map[string]ref)
+
+	// First walk the input to gather all values with a variable reference.
+	_, err = dyn.Walk(r.in, func(p dyn.Path, v dyn.Value) (dyn.Value, error) {
+		ref, ok := newRef(v)
+		if !ok {
+			// Skip values without variable references.
+			return v, nil
+		}
+
+		r.refs[p.String()] = ref
+		return v, nil
+	})
+
+	return err
+}
+
+func (r *resolver) resolveVariableReferences() (err error) {
+	// Initialize map for resolved variables.
+	// We use this for memoization.
+	r.resolved = make(map[string]dyn.Value)
+
+	// Resolve each variable reference (in order).
```

Comment:
Why order matters?

Reply:
Copied this from the `interpolation` package; I recall an issue where the errors that cycle detection produces were non-deterministic because map iteration is random in Go and this means we can enter the cycle at any point in the cycle. By sorting the keys before iterating this will be stable.

Reply:
I'll add this as comment.